docs/en/observability/cloud-monitoring/azure/monitor-azure-openai-apm.asciidoc

```diff
@@ -0,0 +1,203 @@
+
+[source,python]
+----
+
+from openai import AzureOpenAI
+import openai
+from flask import Flask
+import monitor  # Import the module
+from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
+import urllib
+import os
+
+from opentelemetry import trace
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+from opentelemetry.instrumentation.requests import RequestsInstrumentor
+
+# Service name is required for most backends
+resource = Resource(attributes={
+    SERVICE_NAME: "your-service-name"
+})
+
+provider = TracerProvider(resource=resource)
+processor = BatchSpanProcessor(OTLPSpanExporter(endpoint=os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT'),
+        headers="Authorization=Bearer%20"+os.getenv('OTEL_EXPORTER_OTLP_AUTH_HEADER')))
+
+provider.add_span_processor(processor)
+trace.set_tracer_provider(provider)
+tracer = trace.get_tracer(__name__)
+RequestsInstrumentor().instrument()
+
+
+
+# Initialize Flask app and instrument it
+app = Flask(__name__)
+
+# Set OpenAI API key
+client = AzureOpenAI(
+    api_key=os.getenv("AZURE_OPENAI_API_KEY"),
+    api_version=os.getenv("OPENAI_API_VERSION"),
+    azure_endpoint=os.getenv("AZURE_OPENAI_ENDPOINT")
+)
+
+@app.route("/completion")
+@tracer.start_as_current_span("do_work")
+def completion():
+    response = openai.chat.completions.create(
+        model="gpt-4",
+        messages=[
+            {"role": "user", "content": "How do I send my APM data to Elastic Observability?"}
+        ],
+        max_tokens=20,
+        temperature=0
+    )
+
+    return(response.choices[0].message.content.strip())
+
+if __name__ == "__main__":
+    app.run(host="localhost", port=8000, debug=True)
+----
+
+The https://github.com/mdbirnstiehl/AzureOpenAIAPMmonitoringOtel/blob/main/monitor.py[`monitor.py` file] in the example application instruments the application and can be used to instrument your own applications.
+
+The `monitor.py` code uses monkey patching, a technique in Python that dynamically modifies the behavior of a class or module at runtime by modifying its attributes or methods, to modify the behavior of the `chat.completions` call so we can add the response metrics to the OpenTelemetry spans:
+
+[source,python]
+----
+def count_completion_requests_and_tokens(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        counters['completion_count'] += 1
+        response = func(*args, **kwargs)
+
+        token_count = response.usage.total_tokens
+        prompt_tokens = response.usage.prompt_tokens
+        completion_tokens = response.usage.completion_tokens
+        cost = calculate_cost(response)
+        strResponse = json.dumps(response, default=str)
+
+        # Set OpenTelemetry attributes
+        span = trace.get_current_span()
+        if span:
+            span.set_attribute("completion_count", counters['completion_count'])
```

Comment:
What about also storing the messages? Or at least the latest user message? I think our customers want insights into, which types of questions cost more money/tokens. And they can only determine that if they can see the input (prompt).

Another question: are there any upper limit in span attributes that users need to be aware of? Both input messages and responses from the LLM could get very large.

Reply:
I'm afraid the answers to these questions and configuring the code to do these things are above my capabilities. I've basically just taken the code from this [blog](https://www.elastic.co/search-labs/blog/monitor-openai-api-gpt-models-opentelemetry-elastic) and used my very limited abilities to make it work for Azure OpenAI. @vinaychandrasekhar I'm not sure who would be better to answer this.